compiler/erlang_check.erl

```diff
@@ -279,6 +275,113 @@ check_module(File) ->
     end.
 
 %%------------------------------------------------------------------------------
+%% @doc Check for some known files and try to guess what build system is being
+%% used.
+%% @end
+%%------------------------------------------------------------------------------
+-spec guess_build_system(string()) -> {atom(), string()}.
+guess_build_system(Path) ->
+    BuildSystems = [
+                    {rebar, [
+                             "rebar.config",
+                             "rebar.config.script"
+                            ]
+                    },
+                    {make, [
+                            "Makefile"
+                           ]
+                    }
+                   ],
+    case guess_build_system(Path, BuildSystems) of
+        {_, []} -> unknown_build_system;
+        {BuildSystem, Files} -> {BuildSystem, Files}
+    end.
+
+guess_build_system(Path, [{BuildSystem, Files}|Rest]) ->
+    log("Try build system: ~p ~p~n", [BuildSystem]),
+    case find_files(Path, Files) of
+        [] -> guess_build_system(Path, Rest);
+        FoundFiles when is_list(FoundFiles) -> {BuildSystem, FoundFiles}
+    end.
+
+find_files(Path, Files) ->
+    find_files(Path, Files, Files).
+
+find_files("/", _Files, _AllFiles) ->
+    [];
+find_files(_Path, [], _AllFiles) ->
+    [];
+find_files(Path, [File|Rest], AllFiles) ->
+    AbsFile = absname(Path, File),
+    case filelib:is_regular(AbsFile) of
+        true -> log("Found build file: ~p~n", [AbsFile]), [AbsFile];
+        false -> [] ++ find_files(Path, Rest, AllFiles)
+    end ++ find_files(filename:dirname(Path), AllFiles, AllFiles).
+
+load_rebar_files([]) ->
+    error;
+load_rebar_files([ConfigFile|_Rest]) ->
+    ConfigResult = case filename:extension(ConfigFile) of
+                       ".script" -> file:script(ConfigFile);
+                       ".config" -> file:consult(ConfigFile)
+                   end,
+    case ConfigResult of
+        {ok, ConfigTerms} ->
+            log("rebar.config read: ~s~n", [ConfigFile]),
+            {ok, process_rebar_config(filename:dirname(ConfigFile), ConfigTerms)};
```

Comment:
This function should be recursive and call itself; that's how the old `process_rebar_configs` function worked. The reason is that when we have multiple levels of rebar files, rebar uses the options in *all* of them to compile the modules, so we need to do the same.